### PR TITLE
p2p: Capabilities Advertisement for Archival/Catchpoints

### PR DESCRIPTION
**Pull request by Eric-Warehime**
## Summary

PR is currently WIP. There are a few bits which are still being implemented.

This adds libp2p-kad-dht to go-algorand. In this PR it is wrapped in a [cache](https://github.com/libp2p/go-libp2p/blob/master/p2p/discovery/backoff/backoffcache.go) and has some [backoff logic](https://github.com/libp2p/go-libp2p/blob/master/p2p/discovery/backoff/backoff.go#L183) included. 

At a high level it is using the [Content Provider interfaces](https://github.com/libp2p/specs/tree/master/kad-dht#content-provider-advertisement-and-discovery) in the DHT spec in a basic way. It sets up common namespaces for 2 basic capabilities (archival and catchpoints), and advertises those capabilities to bootstrap peers on node startup.

## Next Steps (Not in this PR)

* Integrate the peer store and provider stores with the catchup and fast catchup services in order to use them as the source of peer info (with SRV entries as a fallback).

* In order for non-relay nodes to be able to use the DHT for capabilities advertisement, we'll need the relays to enable it--since the relays in the SRV list will be used as initial peers.

## Test Plan

More testing to be added as the PR progresses. I'll update here.

### CODE REVIEW

**Comment by Eric-Warehime** [comment]
Still need to add testing, merge from master, add some docs, etc. But feel free to give early feedback at this point if anyone has some.

File: network/p2p/capabilities.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+// Host exposes the underlying libp2p host.Host object
+func (c *CapabilitiesDiscovery) Host() host.Host {
+	return c.dht.Host()
+}
+
+// AddPeer adds a given peer.AddrInfo to the Host's Peerstore, and the DHT's routing table
+func (c *CapabilitiesDiscovery) AddPeer(p peer.AddrInfo) (bool, error) {
+	c.Host().Peerstore().AddAddrs(p.ID, p.Addrs, libpeerstore.TempAddrTTL)
```
Comment (shiqizng) [inline network/p2p/capabilities.go:80]: nit: maybe use libp2p.AddressTTL to be consistent with what's used in peerstore initialization?

File: network/p2p/capabilities.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), operationTimeout)
+	defer cancel()
+	var peers []peer.AddrInfo
+	peersChan, err := c.FindPeers(ctx, string(capability), discovery.Limit(n))
+	if err != nil {
+		return nil, err
+	}
+	for p := range peersChan {
+		if p.ID.Size() > 0 {
+			peers = append(peers, p)
```
Comment (shiqizng) [inline network/p2p/capabilities.go:97]: do we need to check that p.ID != node.peerID? does FindPeers return addrInfo for the host node?

File: network/p2p/capabilities.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+		ticker := time.NewTicker(time.Second / 10000)
+		defer func() {
+			ticker.Stop()
+			c.wg.Done()
+		}()
+
+		for {
+			select {
+			case <-c.dht.Context().Done():
+				c.wg.Done()
```
Comment (shiqizng) [inline network/p2p/capabilities.go:122]: is this needed? it's also called in defer func()

File: config/config.go [hunk 4]
```diff
@@ -89,6 +89,10 @@ const MaxGenesisIDLen = 128
 // MaxEvalDeltaTotalLogSize is the maximum size of the sum of all log sizes in a single eval delta.
 const MaxEvalDeltaTotalLogSize = 1024
 
+// ForceCatchpointFileGenerationTrackingMode defines the CatchpointTracking mode that would be used to
+// force a node to generate catchpoint files.
+const ForceCatchpointFileGenerationTrackingMode = 99
```
Comment (winder) [inline config/config.go:94]: Could this refactoring be split into a separate PR?

File: node/node.go [hunk 5]
```diff
@@ -430,6 +459,10 @@ func (node *AlgorandFullNode) Stop() {
 	node.lowPriorityCryptoVerificationPool.Shutdown()
 	node.cryptoPool.Shutdown()
 	node.cancelCtx()
+	node.ledger.Close()
```
Comment (algorandskiy) [inline node/node.go:462]: Closing ledger is a gamble now, we have a [WIP task](https://github.com/algorand/go-algorand-internal/issues/3051) to fix this

File: node/node.go [hunk 6]
```diff
@@ -152,6 +156,8 @@ type AlgorandFullNode struct {
 	tracer messagetracer.MessageTracer
 
 	stateProofWorker *stateproof.Worker
+
+	capabilitiesDiscovery *p2p.CapabilitiesDiscovery
```
Comment (algorandskiy) [inline node/node.go:160]: is it possible to have node free of p2p dependency/implementation detail?

File: ledger/catchpointtracker.go [hunk 7]
```diff
@@ -163,32 +163,9 @@ func (ct *catchpointTracker) initialize(cfg config.Local, paths DirsAndPrefix) {
...
-			ct.catchpointInterval = cfg.CatchpointInterval
-			ct.enableGeneratingCatchpointFiles = true
-		}
-	case forceCatchpointFileGenerationTrackingMode:
-		if cfg.CatchpointInterval > 0 {
-			ct.catchpointInterval = cfg.CatchpointInterval
-			ct.enableGeneratingCatchpointFiles = true
+	if cfg.StoresCatchpoints() {
+		ct.catchpointInterval = cfg.CatchpointInterval
+		if cfg.CatchpointTracking == config.ForceCatchpointFileGenerationTrackingMode {
```
[the remark this answers is not in the document]
Comment (Eric-Warehime) [reply ledger/catchpointtracker.go:168]: Addressed in https://github.com/algorand/go-algorand/pull/5725

File: network/p2p/capabilities.go (hunk 2 above)
Comment (Eric-Warehime) [reply network/p2p/capabilities.go:97]: I had assumed this would be done for us, but I've gone back and looked through the code and I can't find anywhere that they would filter that out. I'll add a test to see what happens and remove the listen address that the Host is listening on from the returned list if it ends up being possible since I think you're right--we don't really ever want to get ourselves in this function call.

Comment (Eric-Warehime) [reply network/p2p/capabilities.go:97]: Done--the new test indeed found itself before I updated this.

File: config/localTemplate.go [hunk 8]
```diff
@@ -672,3 +676,32 @@ func (cfg *Local) AdjustConnectionLimits(requiredFDs, maxFDs uint64) bool {
...
+			return true
+		}
+	case 2:
+		if cfg.CatchpointInterval > 0 {
+			return true
+		}
+	case ForceCatchpointFileGenerationTrackingMode:
+		if cfg.CatchpointInterval > 0 {
+			return true
+		}
```
[the remark this answers is not in the document]
Comment (Eric-Warehime) [reply config/localTemplate.go:883]: It automatically hits the bottom and returns false. I've added tests to ensure it. This was moved to https://github.com/algorand/go-algorand/pull/5725 and I'm rebasing on top of that here now.

File: network/p2p/capabilities.go (hunk 3 above)
Comment (Eric-Warehime) [reply network/p2p/capabilities.go]: Yes it did cause errors when I added tests around it--I've removed it.
